### DISCUSSION TITLE
Combining snap and push jobs

### DISCUSSION THREAD

**Discussion by sr** [zrepl/zrepl#576] [Q&A] [answered]
I've got two `push` jobs on the same host, both replicating all filesystems, one to host A, the other to host B. The two push jobs are exactly the same, except for their `connect.address`.

Initially I had the two jobs configured to perform snapshotting (and sender-side pruning) at interval using `snapshotting.periodic`. However since both jobs would wake up at the same time, they would sometimes stomp on each others, resulting in ZFS errors like "snapshot zrepl_2022_XXX already exists".

To solve this, I moved snapshotting and sender-side pruning to a separate `snap` job, but now I need a way to wake up the two replication jobs (`push` jobs) whenever the `snap` job completes. Hooks seem to be the way to go here, the `snap` job could run `zrepl signal wakeup push1` and `zrepl signal wakeup push2` post snapshot.

Unfortunately, based on the documentation, hooks are executed for each filesystem, so replication would trigger many times during snapshotting, rather than just once, after *all* the snapshots have been created... but maybe that's fine?

**Comment by problame** [comment, marked as answer]
The current hook types aren't a great fit, I'm afraid.
For the time being, you could write a script that monitors the snap job's state (`zrepl status --raw`) and uses `zrepl signal wakeup JOB` on the replication jobs whenever it observes that the snap job finished taking snapshots.

If you want to stay updated on this kind of issue, I suggest you subscribe to https://github.com/zrepl/zrepl/issues/397#issuecomment-766143183

**Reply by sr** [reply]
Thanks, subscribed.

I ended up using a systemd timer to kick off replication at night. I am replicating across two households via Tailscale, so limiting CPU and bandwidth use during the day is important for my particular use case.

```
$ cat /etc/systemd/system/zrepl-wakeup@.timer 
[Unit]
Requires=zrepl-wakeup@%i.service

[Timer]
OnCalendar=*-*-* 3:00:00

[Install]
WantedBy=timers.target
$ cat /etc/systemd/system/zrepl-wakeup@.service
[Unit]
Description=Wake up zrepl job %I
After=zrepl.service
ConditionFileIsExecutable=/usr/bin/zrepl

[Service]
Type=oneshot
ExecStart=/usr/bin/zrepl signal wakeup %I
```

The proposed cron-like trigger would work great.
